### PR TITLE
chore: create kubefedcluster cr for cluster to register itself as host and member

### PR DESCRIPTION
**Pull request by dipak-pawar**

### CODE REVIEW

File: scripts/create_kubefedcluster_cr.sh [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+# getting sa with required secrets and certs
+echo "Switching to project ${NS_TO_GET_SECRET}"
+oc project "${NS_TO_GET_SECRET}"
+
+echo "Getting SA token from ${JOINING_CLUSTER_TYPE} cluster"
+SA_TOKEN=`oc sa get-token ${SA_NAME}`
+SA_SECRET=`oc get sa ${SA_NAME} -o json | jq -r .secrets[].name | grep token`
+SA_CA_CRT=`oc get secret ${SA_SECRET} -o json | jq -r '.data["ca.crt"]'`
+
+# infrastructure CRD is only available in Opeshift 4.x
```
Comment (xcoulon) [inline scripts/create_kubefedcluster_cr.sh]: minor typo: `Opeshift` -> `OpenShift`

File: scripts/create_kubefedcluster_cr.sh [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  NS_TO_CREATE_SECRET="${MEMBER_OPERATOR_NS}"
+fi
+
+# getting sa with required secrets and certs
+echo "Switching to project ${NS_TO_GET_SECRET}"
+oc project "${NS_TO_GET_SECRET}"
+
+echo "Getting SA token from ${JOINING_CLUSTER_TYPE} cluster"
+SA_TOKEN=`oc sa get-token ${SA_NAME}`
+SA_SECRET=`oc get sa ${SA_NAME} -o json | jq -r .secrets[].name | grep token`
```
Comment (xcoulon) [inline scripts/create_kubefedcluster_cr.sh]: what happens if you have multiple entries in the `secrets` array?

File: scripts/create_kubefedcluster_cr.sh [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+SA_CA_CRT=`oc get secret ${SA_SECRET} -o json | jq -r '.data["ca.crt"]'`
+
+# infrastructure CRD is only available in Opeshift 4.x
+API_ENDPOINT=`oc get infrastructure cluster --template={{.status.apiServerURL}}`
+JOINING_CLUSTER_NAME=`oc get infrastructure cluster --template={{.status.infrastructureName}}`
+
+echo "Switching to project ${NS_TO_CREATE_SECRET}"
+oc project "${NS_TO_CREATE_SECRET}"
```
Comment (xcoulon) [inline scripts/create_kubefedcluster_cr.sh]: project is assumed to exist? what happens if it's not the case? (is it even possible it does not exist?)

File: scripts/create_kubefedcluster_cr.sh [hunk 4]
```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/env bash
+
+user_help () {
+    echo "Creates KubeFedCluster"
+    echo "options:"
+    echo "-t, --type            joining cluster type (host or member)"
+    echo "-mn, --member-ns      namespace where member-operator is running"
+    echo "-hn, --host-ns        namespace where host-operator is running"
```
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply scripts/create_kubefedcluster_cr.sh]: we'll using the same script in our end-to-end tests where namespace will be random

File: scripts/create_kubefedcluster_cr.sh [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                JOINING_CLUSTER_TYPE=$1
+                shift
+                ;;
+            -mn|--member-ns)
+                shift
+                if test $# -gt 0; then
+                    MEMBER_OPERATOR_NS=$1
+                    INSTALL_SPECIFIC_VERSION="1"
+                else
+                    echo "Member Operator Namespace is not specified."
```
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply scripts/create_kubefedcluster_cr.sh]: good point

File: scripts/create_kubefedcluster_cr.sh [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+echo "Joining Cluster Type: ${JOINING_CLUSTER_TYPE}"
+echo "Member Operator Namespace: ${MEMBER_OPERATOR_NS}"
+echo "Host Operator Namespace: ${HOST_OPERATOR_NS}"
+
+SA_NAME=${JOINING_CLUSTER_TYPE}"-operator"
+
+# find ns to get secret and create reference secret
+CLUSTER_JOIN_TO="host"
+NS_TO_GET_SECRET="${MEMBER_OPERATOR_NS}"
+NS_TO_CREATE_SECRET="${HOST_OPERATOR_NS}"
```
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply scripts/create_kubefedcluster_cr.sh]: I have updated this with existing script

File: scripts/create_kubefedcluster_cr.sh (hunk 3 above)
Comment (dipak-pawar) [reply scripts/create_kubefedcluster_cr.sh]: yes this script assumes we have projects already created upfront

File: scripts/create_kubefedcluster_cr.sh (hunk 2 above)
Comment (dipak-pawar) [reply scripts/create_kubefedcluster_cr.sh]: we already have multiple entries, this is why we are doing grep on token.

File: scripts/add-cluster.sh [hunk 7]
```diff
@@ -1,44 +1,108 @@
 #!/usr/bin/env bash
 
+user_help () {
+    echo "Creates KubeFedCluster"
+    echo "options:"
+    echo "-t, --type            joining cluster type (host or member)"
+    echo "-mn, --member-ns      namespace where member-operator is running"
+    echo "-hn, --host-ns        namespace where host-operator is running"
+    echo "-p,  --profile        true if using minishift profiles, by default true"
```
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply scripts/add-cluster.sh]: good point, let me fix it

[the remark this answers is not in the document]
Comment (dipak-pawar) [reply scripts/add-cluster.sh]: addressed in  97e90d6

File: scripts/add-cluster.sh [hunk 8]
```diff
@@ -1,44 +1,108 @@
...
+  OPERATOR_NS=${HOST_OPERATOR_NS}
+  CLUSTER_JOIN_TO_OPERATOR_NS=${MEMBER_OPERATOR_NS}
 fi
 
-echo "Switching to profile ${JOINING_CLUSTER_TYPE}"
-minishift profile set ${JOINING_CLUSTER_TYPE}
-oc login -u=system:admin
-oc project "toolchain-${JOINING_CLUSTER_TYPE}-operator"
+# This is using default values i.e. toolchain-member-operator or toolchain-host-operator for local setup
+if [[ ${OPERATOR_NS} == "" &&  ${CLUSTER_JOIN_TO_OPERATOR_NS} == "" ]]; then
```
[the remark this answers is not in the document]
Comment (dipak-pawar) [reply scripts/add-cluster.sh]: added comment in documentation 97e90d6

File: scripts/add-cluster.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply scripts/add-cluster.sh]: Nice, thanks :-) But don't forget to change all parts of the README ;-)

File: README.adoc [hunk 9]
```diff
@@ -114,20 +114,106 @@ $ oc apply -f deploy/user.yaml
...
+    . takes API endpoint and cluster name of the `host` cluster from Kube config
+    . goes to `member` profile
+    . takes cluster name of the `member` cluster from Kube config
+    . creates a secret with the SA token taken from the `host`
+    . creates `KubeFedCluster` CR representing the added `host`
 
-When the script is executed with parameters `add-cluster.sh member` then it does these steps:
+Note: We have make target available for this i.e. `make add-host-to-member` member-operator, host-operator repository
+
+==== Using minishift without profiles or any single openshift cluster as host and member clusters
```
Comment (MatousJobanek) [inline README.adoc]:
```suggestion
==== Using a single cluster for both operators host and member
```

**Pull request merged**